server/admin/webserver/webserver.go

```diff
@@ -0,0 +1,180 @@
+	var authSHA [32]byte
+	fmt.Println("Enter admin webserver password:")
+	password, err := terminal.ReadPassword(syscall.Stdin)
+	if err != nil {
+		return authSHA, err
+	}
+	if password == nil {
+		return authSHA, errors.New("admin webserver password must be set to use the admin webserver")
+	}
+	authSHA = sha256.Sum256(password)
+	// Zero password bytes.
+	for i := range password {
+		password[i] = 0x00
+	}
+	return authSHA, nil
+}
+
+// filesExists reports whether the named file or directory exists.
+func fileExists(name string) bool {
+	_, err := os.Stat(name)
+	return !os.IsNotExist(err)
+}
+
+// New is the constructor for a new WebServer.
+func New(cfg *Config) (*WebServer, error) {
+	// Find the key pair.
+	if !fileExists(cfg.Key) || !fileExists(cfg.Cert) {
+		return nil, fmt.Errorf("missing certificates")
+	}
+
+	keypair, err := tls.LoadX509KeyPair(cfg.Cert, cfg.Key)
+	if err != nil {
+		return nil, err
+	}
+
+	// Prepare the TLS configuration.
+	tlsConfig := &tls.Config{
+		Certificates: []tls.Certificate{keypair},
+		MinVersion:   tls.VersionTLS12,
+	}
+
+	// Create an HTTP router.
+	mux := chi.NewRouter()
+	httpServer := &http.Server{
+		Handler:      mux,
+		ReadTimeout:  rpcTimeoutSeconds * time.Second, // slow requests should not hold connections opened
+		WriteTimeout: rpcTimeoutSeconds * time.Second, // hung responses must die
+	}
+
+	// Make the server.
+	s := &WebServer{
+		core:      cfg.Core,
+		srv:       httpServer,
+		addr:      cfg.Addr,
+		tlsConfig: tlsConfig,
+		authSHA:   cfg.AuthSHA,
+	}
+
+	// Middleware
+	mux.Use(middleware.Recoverer)
+	mux.Use(middleware.RealIP)
+	mux.Use(s.authMiddleware)
+
+	// api endpoints
+	mux.Route("/api", func(r chi.Router) {
+		r.Use(middleware.AllowContentType("application/json"))
+		r.Get("/ping", s.apiPing)
+	})
+
+	return s, nil
+}
+
+// Run starts the web server.
+func (s *WebServer) Run(ctx context.Context) {
+	// Create listener.
+	listener, err := tls.Listen("tcp", s.addr, s.tlsConfig)
+	if err != nil {
+		log.Errorf("can't listen on %s. admin webserver quitting: %v", s.addr, err)
+		return
+	}
+
+	// Close the listener on context cancellation.
+	shutdown := make(chan struct{})
+	go func() {
+		defer close(shutdown)
+		<-ctx.Done()
+
+		if err := s.srv.Shutdown(context.Background()); err != nil {
+			// Error from closing listeners:
+			log.Errorf("HTTP server Shutdown: %v", err)
+		}
+	}()
+	log.Infof("admin webserver listening on %s", s.addr)
+	if err := s.srv.Serve(listener); err != http.ErrServerClosed {
+		log.Warnf("unexpected (http.Server).Serve error: %v", err)
+	}
+
+	// Wait for Shutdown.
+	<-shutdown
+	log.Infof("admin webserver off")
+}
+
+// authMiddleware checks incoming requests for authentication.
+func (s *WebServer) authMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		user, pass, ok := r.BasicAuth()
+		if !ok || s.authSHA != sha256.Sum256([]byte(pass)) {
```

Comment:
Since the admin port should not be open to the public, it probably doesn't matter much, but we may wish to use `subtle.ConstantTimeCompare` here.

Reply:
It's probably also not as important here because we are comparing hashes of the password, but still a good idea.